### PR TITLE
Conflict resolutions conflict

### PR DESCRIPTION
**Pull request by Corvette653**
## AI-Generated Description

# Conflict Resolution Enhancements and Refactor

This PR improves the conflict resolution mechanism in the parser by introducing cycle detection and enhancing error reporting. Additionally, it refactors error handling, optimizes recursive logic for better performance, and strengthens type definitions to improve code maintainability and robustness.

## Key Changes

1. **Cycle Detection in Conflict Resolution**:
   - Added logic to detect and handle cycles during the conflict resolution process.
   - Improved error messages to provide more detailed and actionable feedback.

2. **Refactored Error Handling**:
   - Consolidated error-handling logic for clarity and maintainability.
   - Enhanced exception definitions for better specificity and debugging.

3. **Tail Recursion Optimization**:
   - Refactored `verifyNoConflicts` to leverage tail recursion, improving performance for deep recursion scenarios.

4. **Test Enhancements**:
   - Added new test cases in `ParseTableTest` to validate conflict resolution and cycle detection.

5. **Minor Adjustments**:
   - Updated type definitions for improved clarity and type safety.
   - Small fixes in `createTables` for consistency.

## Technical Details

- **Files Updated**:
  - `ConflictException.scala`: Added new exception handling for detailed conflict reports (+18 lines).
  - `ConflictResolution.scala`: Implemented cycle detection and refactored recursive logic (+41/-1 lines).
  - `createTables.scala`: Minor adjustments for improved readability and consistency (+4/-3 lines).
  - `ParseTableTest.scala`: Expanded test coverage for the new conflict resolution logic (+24 lines).

These changes aim to ensure the parser is more resilient, easier to debug, and performs better in edge cases involving complex conflict scenarios.

---

### Statistics

- **4** files changed
- **87** additions
- **4** deletions

### CODE REVIEW

File: src/alpaca/internal/parser/ConflictException.scala [hunk 1]
```diff
@@ -53,3 +53,19 @@ final class ReduceReduceConflict(red1: Reduction, red2: Reduction, path: List[Sy
...
+  case Production.NonEmpty(lhs, rhs, null) => show"Reduction(${rhs.mkShow(" ")} -> $lhs)"
+  case Production.Empty(lhs, null) => show"Reduction(${Symbol.Empty} -> $lhs)"
+  case p: Production => show"Reduction(${p.name})"
+  case s: String => show"Shift($s)"
+
+final class InconsistentConflictResolution(node: ConflictKey, path: List[ConflictKey])
+  extends ConflictException(
+    show"""
+          |Inconsistent conflict resolution detected:
+          |${path.dropWhile(_ != node).map(show).mkShow(" before ")} before ${show(node)}
```
Comment (halotukozak) [inline src/alpaca/internal/parser/ConflictException.scala]: `.map(show)` should not be required if `given Showable[ConflictKey]` defined

File: src/alpaca/internal/parser/ConflictResolution.scala [hunk 2]
```diff
@@ -66,6 +67,27 @@ private[parser] object ConflictResolutionTable {
       winsOver(first, second) orElse winsOver(second, first)
     }
 
+    def verifyNoConflicts: Unit = {
```
Comment (halotukozak) [inline src/alpaca/internal/parser/ConflictResolution.scala]: Method with Unit result type is parameterless. It performs side effects so should be 
`def verifyNoConflicts()`

**Review by halotukozak — CHANGES_REQUESTED** [review]
Mówisz, masz

File: src/alpaca/internal/parser/ConflictResolution.scala [hunk 3]
```diff
@@ -66,6 +67,26 @@ private[parser] object ConflictResolutionTable {
       winsOver(first, second) orElse winsOver(second, first)
     }
 
+    def verifyNoConflicts() = {
```
Comment (halotukozak) [inline src/alpaca/internal/parser/ConflictResolution.scala]: Type annotation required for Unit definition

File: src/alpaca/internal/parser/ConflictException.scala [hunk 4]
```diff
@@ -53,3 +53,19 @@ final class ReduceReduceConflict(red1: Reduction, red2: Reduction, path: List[Sy
           |Consider marking one of the productions to be alwaysBefore or alwaysAfter the other
           |""".stripMargin,
   )
+
+def show(x: ConflictKey): String = x match
+  case Production.NonEmpty(lhs, rhs, null) => show"Reduction(${rhs.mkShow(" ")} -> $lhs)"
```
[the remark this answers is not in the document]
Comment (halotukozak) [reply src/alpaca/internal/parser/ConflictException.scala]: OK, I've noticed that you reverse -> here. Hence your preference?

File: src/alpaca/internal/parser/createTables.scala [hunk 5]
```diff
@@ -186,6 +186,7 @@ private def createTablesImpl[Ctx <: ParserCtx: Type](
             case Some(set) => Some(set + extractKey(after))
             case None => Some(Set(extractKey(after))),
   ).tap: table =>
+    table.verifyNoConflicts
```
Comment (halotukozak) [inline src/alpaca/internal/parser/createTables.scala]: this line doesn't compile

File: src/alpaca/internal/parser/ConflictResolution.scala [hunk 6]
```diff
@@ -66,6 +67,26 @@ private[parser] object ConflictResolutionTable {
...
+      enum VisitState:
```
Comment (halotukozak) [inline src/alpaca/internal/parser/ConflictResolution.scala]: short enums can be simplified to
```scala
enum VisitState:
        case Unvisited, Visited, Processed
```

File: src/alpaca/internal/parser/ConflictResolution.scala [hunk 7]
```diff
@@ -66,6 +67,26 @@ private[parser] object ConflictResolutionTable {
...
+      def visit(node: ConflictKey, path: List[ConflictKey] = Nil): Unit = visited(node) match
+        case VisitState.Unvisited =>
+          visited.update(node, VisitState.Visited)
+          for neighbor <- table.getOrElse(node, Set.empty) do visit(neighbor, node :: path)
+          visited.update(node, VisitState.Processed)
+        case VisitState.Visited =>
+          throw InconsistentConflictResolution(node, path.reverse)
+        case VisitState.Processed => // Already fully processed
+
+      for node <- table.keys do visit(node)
```
Comment (halotukozak) [inline src/alpaca/internal/parser/ConflictResolution.scala]: (not sure)
can't we filter the pipeline?
```scala
for
        node <- table.keys
        if visited(node) == VisitState.Unvisited
      do
```

File: src/alpaca/internal/parser/ConflictResolution.scala [hunk 8]
```diff
@@ -66,6 +67,27 @@ private[parser] object ConflictResolutionTable {
...
 
+    def verifyNoConflicts: Unit = {
+      enum VisitState:
+        case Unvisited
+        case Visited
+        case Processed
+
+      val visited = mutable.Map.empty[ConflictKey, VisitState].withDefaultValue(VisitState.Unvisited)
+
+      def visit(node: ConflictKey, path: List[ConflictKey] = Nil): Unit =
```
[the remark this answers is not in the document]
Comment (halotukozak) [reply src/alpaca/internal/parser/ConflictResolution.scala]: something like this one
```scala
enum Action:
  case Enter(node: ConflictKey, path: List[ConflictKey])
  case Leave(node: ConflictKey)

@tailrec
def loop(stack: List[Action]): Unit = stack match
  case Nil =>
  
  case Action.Leave(node) :: rest =>
    visited(node) = VisitState.Processed
    loop(rest)
  
  case Action.Enter(node, path) :: rest =>
    visited(node) match
      case VisitState.Processed =>
        loop(rest)
  
      case VisitState.Visited =>
        throw InconsistentConflictResolution(node, path.reverse)
  
      case VisitState.Unvisited =>
        visited(node) = VisitState.Visited
  
        val nextActions = table
          .getOrElse(node, Set.empty)
          .map(Action.Enter(_, node :: path))
          .toList
  
        loop(nextActions ::: (Action.Leave(node) :: rest))
```

[the remark this answers is not in the document]
Comment (halotukozak) [reply src/alpaca/internal/parser/ConflictResolution.scala]: you don't want new enum? no problem
```scala
def verifyNoConflicts(): Unit = {
  enum VisitState:
    case Visited, Processed, Unvisited

  val visited = mutable.Map.empty[ConflictKey, VisitState].withDefaultValue(VisitState.Unvisited)

  @tailrec
  def loop(stack: List[(node: ConflictKey, path: Option[List[ConflictKey]])]): Unit = stack match
    case Nil =>

    case (node, None) :: rest =>
      visited(node) = VisitState.Processed
      loop(rest)

    case (node, Some(path)) :: rest =>
      visited(node) match
        case VisitState.Processed =>
          loop(rest)

        case VisitState.Visited =>
          throw InconsistentConflictResolution(node, path.reverse)

        case VisitState.Unvisited =>
          visited(node) = VisitState.Visited

          val nextActions = for
            values <- table.get(node).iterator
            neighbor <- values
          yield (neighbor, Some(node :: path))

          loop(nextActions.toList ::: (node, None) :: rest)

  for
    node <- table.keys
    if visited(node) == VisitState.Unvisited
  do loop(List((node, Some(Nil))))
  ```

File: src/alpaca/internal/parser/ConflictException.scala [hunk 9]
```diff
@@ -53,3 +53,21 @@ final class ReduceReduceConflict(red1: Reduction, red2: Reduction, path: List[Sy
           |Consider marking one of the productions to be alwaysBefore or alwaysAfter the other
           |""".stripMargin,
   )
+
+final class InconsistentConflictResolution(node: ConflictKey, path: List[ConflictKey])
+  extends ConflictException(
+    {
+      given Showable[ConflictKey] =
```
[the remark this answers is not in the document]
Comment (halotukozak) [reply src/alpaca/internal/parser/ConflictException.scala]: or even make `ConflictKey` opaque and add some factory

```scala
opaque private[parser] type ConflictKey = Production | String

object ConflictKey:
  inline def apply(key: Production | String): ConflictKey = key
```

then 
1. import won't be necessary
2. union type won't overlap Production / String type classes

File: src/alpaca/internal/parser/ConflictResolution.scala (hunk 3 above)
Comment (Corvette653) [reply src/alpaca/internal/parser/ConflictResolution.scala]: Interesting, I do not get this error

File: src/alpaca/internal/parser/ConflictResolution.scala (hunk 7 above)
Comment (Corvette653) [reply src/alpaca/internal/parser/ConflictResolution.scala]: isn't it the same complexity to run this if vs 
```
visited(node) match
  case VisitState.Processed => // Already fully processed
```
?

**Comment by halotukozak** [comment]
@Corvette653 What about this one?
